### ISSUE TITLE
Geolocate customer using IP address

### ISSUE THREAD

**Issue by dwightkelly** [vendure-ecommerce/vendure#2596]
For a variety of reasons it is desirable to detect the location of a customer. This can be done using a service such as ipstack to geolocate the customer's IP address. It's not fullproof because of VPN/Tor/etla. 

The location can be used to map customer to a region, location based promotions, enhance logging and statistics, disallow access if outside of supported region(s), DDOS protection, etc.

It would also be useful to make the customer's location info available to other Vendure services and plugins.

**Comment by michaelbromley** [comment]
In general I think the option of accessing more specific request information is good, but I don't want to couple Vendure core with a specific paid geolocation service.

One thing I've been thinking of is to make the `RequestContext` (or `Session`) object easily extensible. Then we could have a plugin which enriches it with geolocation (and whatever else) data, which could then be used by other parts of the system.

Related:

- https://github.com/vendure-ecommerce/vendure/issues/2069

**Comment by michaelbromley** [comment]
Yes reverse proxies & load balancers will complicate this, so we can't make a blanket assumption about how to detect the client IP. Any solution would have to be configurable based on prod environment. 

If we allow adding arbitrary data to `RequestContext`, it should come with an API that allows us to assign this at the point that the `RequestContext` object is created: namely our internal `AuthGuard` through which each incoming request will pass. You could then detect the IP based on the known request headers available in your environment, and then attach them to the `ctx` which can then be used throughout the rest of the app for things like promotions & statistic logging.

**Comment by dlhck** [comment]
This feature request has been moved to GitHub Discussions for community voting.
Please upvote the discussion if you would like to see this implemented: https://github.com/vendurehq/vendure/discussions/4274
